From b4f371fbc4c9faf340ae85662274bfb3ad255e19 Mon Sep 17 00:00:00 2001
From: Vinod Koul <vinod.koul@linux.intel.com>
Date: Thu, 21 Jul 2011 18:55:21 +0100
Subject: [PORT FROM MAIN] tinyalsa: add upstream changes required for mrfld

BZ: 66195

This adds two patches one for asound.h addition and second for ctl

1.
include: Add a local asound.h to allow build with current public releases

Bionic does not ship a copy of asound.h and it should do no harm to have
a local copy here.

upstream commit-id: 648e14a421f1795006f770de8c57602366b20c24

2.
Add mixer_ctl_{set,get}_data()

For binary controls we don't want to go through
mixer_ctl_{set,get}_value() as that will trigger many calls to our
get()/put() callback in the kernel.  Set the entire payload at once
and trigger the get()/put() callback once.

This also updated mixer array size to 1024

Change-Id: I6b0815aee114194279100ac9f0cb2ea5968a08ac
Orig-Change-Id: Ic94333ab794299058d149d493d384c301e5d5f25
Signed-off-by: Vinod Koul <vinod.koul@linux.intel.com>
Signed-off-by: Omair Mohammed Abdullah <omair.m.abdullah@linux.intel.com>
Reviewed-on: http://android.intel.com:8080/82296
Reviewed-by: Centelles, Sylvain <sylvain.centelles@intel.com>
Reviewed-by: M, Arulselvan <arulselvan.m@intel.com>
Tested-by: M, Arulselvan <arulselvan.m@intel.com>
Reviewed-by: cactus <cactus@intel.com>
Tested-by: cactus <cactus@intel.com>
---
 include/sound/asound.h       |    2 +-
 include/tinyalsa/asoundlib.h |    2 +
 mixer.c                      |   48 ++++++++++++++++++++++++++++++++++++++++++
 3 files changed, 51 insertions(+), 1 deletions(-)

diff --git a/include/sound/asound.h b/include/sound/asound.h
index a041628..e7402b2 100644
--- a/include/sound/asound.h
+++ b/include/sound/asound.h
@@ -740,7 +740,7 @@ struct snd_ctl_elem_value {
  unsigned int *item_ptr;
  } enumerated;
  union {
- unsigned char data[512];
+ unsigned char data[1024];
  unsigned char *data_ptr;
  } bytes;
  struct snd_aes_iec958 iec958;
diff --git a/include/tinyalsa/asoundlib.h b/include/tinyalsa/asoundlib.h
index 3f31342..6346da4 100644
--- a/include/tinyalsa/asoundlib.h
+++ b/include/tinyalsa/asoundlib.h
@@ -231,8 +231,10 @@ int mixer_ctl_get_percent(struct mixer_ctl *ctl, unsigned int id);
 int mixer_ctl_set_percent(struct mixer_ctl *ctl, unsigned int id, int percent);
 
 int mixer_ctl_get_value(struct mixer_ctl *ctl, unsigned int id);
+int mixer_ctl_get_data(struct mixer_ctl *ctl, void *data, size_t len);
 int mixer_ctl_get_bytes(struct mixer_ctl *ctl, void *data, size_t len);
 int mixer_ctl_set_value(struct mixer_ctl *ctl, unsigned int id, int value);
+int mixer_ctl_set_data(struct mixer_ctl *ctl, const void *data, size_t len);
 int mixer_ctl_set_bytes(struct mixer_ctl *ctl, const void *data, size_t len);
 int mixer_ctl_set_enum_by_string(struct mixer_ctl *ctl, const char *string);
 
diff --git a/mixer.c b/mixer.c
index f52bca9..61055c8 100644
--- a/mixer.c
+++ b/mixer.c
@@ -317,6 +317,32 @@ int mixer_ctl_get_value(struct mixer_ctl *ctl, unsigned int id)
     return 0;
 }
 
+int mixer_ctl_get_data(struct mixer_ctl *ctl, void *data, size_t len)
+{
+    struct snd_ctl_elem_value ev; 
+    int ret;
+
+    if (!ctl || (len > ctl->info->count) || !data)
+        return -EINVAL;
+
+    memset(&ev, 0, sizeof(ev));
+    ev.id.numid = ctl->info->id.numid;
+    ret = ioctl(ctl->mixer->fd, SNDRV_CTL_IOCTL_ELEM_READ, &ev);
+    if (ret < 0)
+        return ret;
+
+    switch (ctl->info->type) {
+    case SNDRV_CTL_ELEM_TYPE_BYTES:
+        memcpy(data, ev.value.bytes.data, len);
+        break;
+
+    default:
+        return -EINVAL;
+    }   
+
+    return 0;
+}
+
 int mixer_ctl_get_bytes(struct mixer_ctl *ctl, void *data, size_t len)
 {
     struct snd_ctl_elem_value ev;
@@ -372,6 +398,28 @@ int mixer_ctl_set_value(struct mixer_ctl *ctl, unsigned int id, int value)
     return ioctl(ctl->mixer->fd, SNDRV_CTL_IOCTL_ELEM_WRITE, &ev);
 }
 
+int mixer_ctl_set_data(struct mixer_ctl *ctl, const void *data, size_t len)
+{
+    struct snd_ctl_elem_value ev;
+
+    if (!ctl || (len > ctl->info->count) || !len || !data)
+        return -EINVAL;
+
+    memset(&ev, 0, sizeof(ev));
+    ev.id.numid = ctl->info->id.numid;
+
+    switch (ctl->info->type) {
+    case SNDRV_CTL_ELEM_TYPE_BYTES:
+        memcpy(ev.value.bytes.data, data, len);
+        break;
+
+    default:
+        return -EINVAL;
+    }
+
+    return ioctl(ctl->mixer->fd, SNDRV_CTL_IOCTL_ELEM_WRITE, &ev);
+}
+
 int mixer_ctl_set_bytes(struct mixer_ctl *ctl, const void *data, size_t len)
 {
     struct snd_ctl_elem_value ev;
-- 
1.7.4.1

